[ARM32/RyuJIT] Use regtype instead of node's type in PUTARG_SPLIT case

jyoungyun opened the pull request:
The size of TYP_STRUCT in definition is 1 and it is lower than sizeof(int).
It means TYP_STRUCT should use the node's type instead of its type.
In PUTARG_SPLIT case, it wil use its regtype of child nodes.

Related issue: #12990 

/cc @dotnet/arm32-contrib

jyoungyun commented:
Even though applying this changes, below testcases still failed with the following errors.

Failed tc:
```
JIT/Methodical/divrem/div/overlddiv_cs_do/overlddiv_cs_do.sh
JIT/Methodical/divrem/div/overlddiv_cs_ro/overlddiv_cs_ro.sh
JIT/Methodical/divrem/rem/overldrem_cs_do/overldrem_cs_do.sh
JIT/Methodical/divrem/rem/overldrem_cs_ro/overldrem_cs_ro.sh
```

New assertion:
```
Assert failure(PID 5095 [0x000013e7], Thread: 5095 [0x13e7]): Assertion failed '!"gtArgEntryByNode: node not found"' in 'overldrem:Main():int' (IL size 28941)

    File: /opt/code/src/jit/gentree.cpp Line: 6488
```

jyoungyun commented:
@dotnet-bot please test Tizen armel Cross Debug Build

hseok-oh commented:
@jyoungyun We can make tmpVar in morphing phase for argument that needs many operation, instead of using GT_RELOAD.

CarolEidt commented:
> Is the GT_RELOAD added under the GT_CALL node the correct behavior?

Yes, the `GT_RELOAD` is required when a node has been spilled. It seems that code generation should be using the `GT_RELOAD` node to search the arg table.

> We can make tmp LclVar in morphing phase for argument that needs many operation, instead of using GT_RELOAD.

While that might be a reasonable thing to do, the code generator must still correctly handle the case where a `GT_RELOAD` occurs above any node that produces a register.

CarolEidt commented:
I believe that you need to do the following to fix the assert you are seeing now:
In codegenarmarch.cpp, in `CodeGen::genCallInstruction()`:
Change this:
```
        fgArgTabEntryPtr curArgTabEntry = compiler->gtArgEntryByNode(call, argNode->gtSkipReloadOrCopy());
```
to this:
```
        fgArgTabEntryPtr curArgTabEntry = compiler->gtArgEntryByNode(call, argNode);
        argNode = argNode->gtSkipReloadOrCopy();
```

jyoungyun commented:
@CarolEidt Yes, I tried to change the code as you said and this assertion was passed and the code generation worked well. In generating code of PUTARG_SPLIT, the register r2 spilled and the node is unspilled in CALL node. (But gc debugging section still have some issue. And I will continue to look it.)

```
                                                              ┌──▌  t3352  void
Generating: N5015 (???,???) [005197] ---XG-------Z     t5197 = ▌  PUTARG_SPLIT struct REG r2
                                                        Tree-Node marked unspilled from  [003351]
IN099a:             ldr     r2, [sp+0x14]       // [TEMP_02]
release temp #2, slot 0, size = 4
                                                        Tree-Node marked unspilled from  [003383]
IN099b:             ldr     lr, [sp+0x10]       // [TEMP_01]
release temp #1, slot 0, size = 4
IN099c:             str     lr, [sp]    // [V110 OutArgs]
IN099d:             str     r0, [sp+0x04]       // [V110 OutArgs+0x04]
reused temp #1, slot 0, size = 4
                                                        The register r2 spilled with    [005197]
IN099e:             str     r2, [sp+0x10]       // [TEMP_01]

                                                              ┌──▌  t6841  struct arg1 r2,r3 out+00
                                                              ├──▌  t5198  ref    arg0 in r0
                                                              ├──▌  t5200  int    control expr
Generating: N5103 (249,174) [002499] --CXG--------     t2499 = ▌  CALL      int    numHolder.op_Division $c35
                                                        Tree-Node marked unspilled from  [005197]
IN09d1:             ldr     r2, [sp+0x10]       // [TEMP_01]
release temp #1, slot 0, size = 4
                                                        GC regs: 03E1 {r0 r5 r6 r7 r8 r9} => 03E0 {r5 r6 r7 r8 r9}
Call: GCvars=0000000000154000000000030000000000000000 {V07 V09 V11 V13 V14 V15}, gcrefRegs=03E0 {r5 r6 r7 r8 r9}, byrefRegs=0000 {}
[2514] Rec call GC vars = 0000000000154000000000030000000000000000
IN09d2:             blx     r2          // numHolder:op_Division(ref,struct):int
```

@hseok-oh And I also tried to use LclVar in morphing phase for argument and it worked well too.
```
N044 (142,108) [002499] --CXG--------                   └──▌  CALL      int    numHolder.op_Division $c2f
N013 ( 46, 31) [002485] ---X---------                      │  ┌──▌  IND       ref    <l:$36c, c:$bae>
N012 ( 43, 29) [002484] ---X---------                      │  │  └──▌  ARR_ELEM[,] byref  $c65
N001 (  1,  1) [002473] -------------                      │  │     ├──▌  LCL_VAR   ref    V01 loc1         u:3 $310
N006 ( 16, 11) [002478] ---X---------                      │  │     ├──▌  IND       int    <l:$5b3, c:$c27>
N005 ( 13,  9) [002477] ---X---------                      │  │     │  └──▌  ARR_ELEM[,] byref  $c63
N002 (  1,  1) [002474] -------------                      │  │     │     ├──▌  LCL_VAR   ref    V16 loc16        u:3 $a7f
N003 (  1,  1) [002475] -------------                      │  │     │     ├──▌  CNS_INT   int    0 $40
N004 (  1,  1) [002476] -------------                      │  │     │     └──▌  CNS_INT   int    1 $42
N011 ( 16, 11) [002483] ---X---------                      │  │     └──▌  IND       int    <l:$5b4, c:$c28>
N010 ( 13,  9) [002482] ---X---------                      │  │        └──▌  ARR_ELEM[,] byref  $c64
N007 (  1,  1) [002479] -------------                      │  │           ├──▌  LCL_VAR   ref    V16 loc16        u:3 $a7f
N008 (  1,  1) [002480] -------------                      │  │           ├──▌  CNS_INT   int    1 $42
N009 (  1,  1) [002481] -------------                      │  │           └──▌  CNS_INT   int    0 $40
N015 ( 50, 34) [003370] -A-X----R--L- arg0 SETUP           ├──▌  ASG       ref    <l:$36c, c:$bae>
N014 (  3,  2) [003369] D------N-----                      │  └──▌  LCL_VAR   ref    V113 tmp96       d:3 <l:$36c, c:$bae>
N028 ( 48, 32) [002500] ---X---N-----                      │  ┌──▌  IND       struct <l:$5de, c:$61b>
N027 ( 45, 30) [002497] ---X---------                      │  │  └──▌  ARR_ELEM[,] byref  $c68
N016 (  3,  2) [002486] -------------                      │  │     ├──▌  LCL_VAR   ref    V14 loc14        u:3 $818
N021 ( 16, 11) [002491] ---X---------                      │  │     ├──▌  IND       int    <l:$5b5, c:$c29>
N020 ( 13,  9) [002490] ---X---------                      │  │     │  └──▌  ARR_ELEM[,] byref  $c66
N017 (  1,  1) [002487] -------------                      │  │     │     ├──▌  LCL_VAR   ref    V16 loc16        u:3 $a7f
N018 (  1,  1) [002488] -------------                      │  │     │     ├──▌  CNS_INT   int    0 $40
N019 (  1,  1) [002489] -------------                      │  │     │     └──▌  CNS_INT   int    1 $42
N026 ( 16, 11) [002496] ---X---------                      │  │     └──▌  IND       int    <l:$5b6, c:$c2a>
N025 ( 13,  9) [002495] ---X---------                      │  │        └──▌  ARR_ELEM[,] byref  $c67
N022 (  1,  1) [002492] -------------                      │  │           ├──▌  LCL_VAR   ref    V16 loc16        u:3 $a7f
N023 (  1,  1) [002493] -------------                      │  │           ├──▌  CNS_INT   int    1 $42
N024 (  1,  1) [002494] -------------                      │  │           └──▌  CNS_INT   int    0 $40
N030 ( 52, 35) [003368] -A-X----R--L- arg1 SETUP           ├──▌  ASG       struct (copy) $VN.Void
N029 (  3,  2) [003367] D------N-----                      │  └──▌  LCL_VAR   struct(AX) V102 tmp85
N033 (  3,  2) [003371] ------------- arg0 in r0           ├──▌  LCL_VAR   ref    V113 tmp96       u:3 (last use) <l:$36c, c:$bae>
N038 (  3,  5) [003384] -------------                      │        ┌──▌  FIELD_LIST void   int at offset 12 $d21
N037 (  3,  4) [003383] -------------                      │        │  └──▌  LCL_FLD   int    V102 tmp85       [+12] $c2e
N039 (  7, 11) [003382] -------------                      │     ┌──▌  FIELD_LIST void   int at offset 8 $d22
N036 (  3,  4) [003381] -------------                      │     │  └──▌  LCL_FLD   int    V102 tmp85       [+8] $c2d
N040 ( 11, 17) [003380] -------------                      │  ┌──▌  FIELD_LIST void   int at offset 4 $d23
N035 (  3,  4) [003379] -------------                      │  │  └──▌  LCL_FLD   int    V102 tmp85       [+4] $c2c
N041 ( 15, 23) [003378] ------------- arg1 r2,r3 out+00         └──▌  FIELD_LIST void   int at offset 0 $d24
N034 (  3,  4) [003377] -------------                         └──▌  LCL_FLD   int    V102 tmp85       [+0] $c2b
```

Both methods can solve this issue. However, both are likely to cause problems later on, so both methods should be merged together. What do you think?

CarolEidt commented:
I think that the change in codegenarmarch.cpp is required in any case.
I'm not absolutely certain where in morph you are adding a lclVar. That *may* result in better code, though it will increase the number of lclVars, which can lead to other issues for very large methods.

The pros of adding a lclVar:
- if it is spilled it can be readily reloaded from its stack location
- it reduces the register pressure around a call with many arguments, which will reduce register allocation anomalies

The cons of adding a lclVar:
- each lclVar you add will always require its own stack slot (always leading to an increased frame size)
- for a very large method, it could cause it to run into the limit on the number of lclVars

Another option to pursue is to make them regOptional. Then you won't ever encounter a `GT_RELOAD`, since it is assumed that it can be used directly from the spill temp location. However, this requires a bit of extra care in handling, because the associated register may be in use right up until the call or the position at which it has a corresponding `RefTypeFixedUse` (at which point it is presumed killed, so can no longer contain any other values).

jyoungyun commented:
@CarolEidt Thank you for comment. I will be more concerned about adding lclVars for struct arguments.

When I changed the code as you said, register allocator does not work well.

Wrong register allocation:
```
                                                              ┌──▌  t5199  int
Generating: N5101 (  6,  9) [005200] -------------     t5200 = ▌  IND       int    REG r2
IN09d0:             ldr     r2, [r2]
                                                              ┌──▌  t6841  struct arg1 r2,r3 out+00
                                                              ├──▌  t5198  ref    arg0 in r0
                                                              ├──▌  t5200  int    control expr
Generating: N5103 (249,174) [002499] --CXG--------     t2499 = ▌  CALL      int    numHolder.op_Division $c35
                                                        Tree-Node marked unspilled from  [005197]
IN09d1:             ldr     r2, [sp+0x10]       // [TEMP_01]
release temp #1, slot 0, size = 4
                                                        GC regs: 03E1 {r0 r5 r6 r7 r8 r9} => 03E0 {r5 r6 r7 r8 r9}
Call: GCvars=0000000000154000000000030000000000000000 {V07 V09 V11 V13 V14 V15}, gcrefRegs=03E0 {r5 r6 r7 r8 r9}, byrefRegs=0000 {}
[2514] Rec call GC vars = 0000000000154000000000030000000000000000
IN09d2:             blx     r2          // numHolder:op_Division(ref,struct):int
```

```
<RefPosition #8136 @5016 RefTypeDef <Ivl:1828> PUTARG_SPLIT BB06 regmask=[r2] fixed>
<RefPosition #8222 @5103 RefTypeUse <Ivl:1828> BB06 regmask=[r0-r9 r12 lr] last>
<RefPosition #8223 @5103 RefTypeUse <Ivl:1829> BB06 regmask=[r0-r9 r12 lr] last>
```

The r2 is already used by t5200 node, but when the t5197 is restored from stack it uses the r2 too. So the function address is lost and occurs segfault.

I think the interval 1828 should be inherited the Def setting but it doesn't.
So I added more code for assigning the matched source candidates register.

Fixed RefPosition:
```
<RefPosition #8136 @5016 RefTypeDef <Ivl:1828> PUTARG_SPLIT BB06 regmask=[r2] fixed>
<RefPosition #8222 @5103 RefTypeFixedReg <Reg:r2 > BB06 regmask=[r2]>
<RefPosition #8223 @5103 RefTypeUse <Ivl:1828> BB06 regmask=[r2] last fixed>
<RefPosition #8224 @5103 RefTypeFixedReg <Reg:r3 > BB06 regmask=[r3]>
<RefPosition #8225 @5103 RefTypeUse <Ivl:1829> BB06 regmask=[r3] last fixed>
```
Fixed codes:
```
                                                              ┌──▌  t5199  int
Generating: N5101 (  6,  9) [005200] -------------     t5200 = ▌  IND       int    REG lr
IN09d0:             ldr     lr, [r2]
                                                              ┌──▌  t5197  struct arg1 r2,r3 out+00
                                                              ├──▌  t5198  ref    arg0 in r0
                                                              ├──▌  t5200  int    control expr
Generating: N5103 (249,174) [002499] --CXG--------     t2499 = ▌  CALL      int    numHolder.op_Division $c35
                                                        Tree-Node marked unspilled from  [005197]
IN09d1:             ldr     r2, [sp+0x10]       // [TEMP_01]
release temp #1, slot 0, size = 4
                                                        GC regs: 03E1 {r0 r5 r6 r7 r8 r9} => 03E0 {r5 r6 r7 r8 r9}
Call: GCvars=0000000000154000000000030000000000000000 {V07 V09 V11 V13 V14 V15}, gcrefRegs=03E0 {r5 r6 r7 r8 r9}, byrefRegs=0000 {}
[2514] Rec call GC vars = 0000000000154000000000030000000000000000
IN09d2:             blx     lr          // numHolder:op_Division(ref,struct):int
```

jyoungyun commented:
@BruceForstall @CarolEidt Please take a look.

/cc @hseok-oh @wateret @hqueue

CarolEidt commented:
> So I added more code for assigning the matched source candidates register.

Yes, the `PUTARG` nodes are unusual in that they reflect a "more" fixed register requirement than most. In fact, it is more critical that the source candidates be set appropriately, because it is the call (which is the one consuming it as a source) which must have it in the right register.

jyoungyun commented:
@CarolEidt Thank you for review!
